Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt

```diff
@@ -0,0 +1,191 @@
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaData
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpacAndroid
+import de.rki.coronawarnapp.storage.LocalData
+import io.mockk.MockKAnnotations
+import io.mockk.Runs
+import io.mockk.clearAllMocks
+import io.mockk.coEvery
+import io.mockk.coVerify
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.just
+import io.mockk.mockkObject
+import io.mockk.spyk
+import org.joda.time.Days
+import org.joda.time.Instant
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import testhelpers.BaseTest
+import testhelpers.coroutines.runBlockingTest2
+import testhelpers.preferences.mockFlowPreference
+
+class AnalyticsTest : BaseTest() {
+    @MockK lateinit var dataDonationAnalyticsServer: DataDonationAnalyticsServer
+    @MockK lateinit var appConfigProvider: AppConfigProvider
+    @MockK lateinit var deviceAttestation: DeviceAttestation
+    @MockK lateinit var settings: AnalyticsSettings
+    @MockK lateinit var configData: ConfigData
+    @MockK lateinit var analyticsConfig: AnalyticsConfig
+    @MockK lateinit var exposureRiskMetadataDonor: ExposureRiskMetadataDonor
+    @MockK lateinit var lastAnalyticsSubmissionLogger: LastAnalyticsSubmissionLogger
+
+    @BeforeEach
+    fun setup() {
+        MockKAnnotations.init(this)
+
+        mockkObject(LocalData)
+
+        coEvery { appConfigProvider.getAppConfig() } returns configData
+        every { configData.analytics } returns analyticsConfig
+
+        coEvery { lastAnalyticsSubmissionLogger.storeAnalyticsData(any()) } just Runs
+    }
+
+    @AfterEach
+    fun tearDown() {
+        clearAllMocks()
+    }
+
+    private fun createDonorModules(): Set<DonorModule> = setOf(exposureRiskMetadataDonor)
+
+    private fun createInstance() = spyk(
+        Analytics(
+            dataDonationAnalyticsServer = dataDonationAnalyticsServer,
+            appConfigProvider = appConfigProvider,
+            deviceAttestation = deviceAttestation,
+            donorModules = createDonorModules(),
+            settings = settings,
+            logger = lastAnalyticsSubmissionLogger
+        )
+    )
+
+    @Test
+    fun `abort due to no user consent`() {
+        every { settings.analyticsEnabled } returns mockFlowPreference(false)
+
+        val analytics = createInstance()
+
+        runBlockingTest2 {
+            analytics.submitIfWanted()
+        }
+
+        coVerify(exactly = 0) {
+            analytics.submitAnalyticsData()
+        }
+    }
+
+    @Test
+    fun `abort due to submit probability`() {
+        every { settings.analyticsEnabled } returns mockFlowPreference(false)
+        every { analyticsConfig.probabilityToSubmit } returns 0.0
+
+        val analytics = createInstance()
+
+        runBlockingTest2 {
+            analytics.submitIfWanted()
+        }
+
+        // TODO: is there a way to make sure a spy function returned x?
```

Comment:
You could inject the randomness source and hardcode it to a specific value.

We currently have `SecureRandom` in the DI graph for SafetyNet.

Reply:
this comment was more in reference to https://github.com/corona-warn-app/cwa-app-android/pull/2280#discussion_r573931976